Improve the Trash data model

johnswanson opened the pull request:
Ok, so I had a realization at the PERFECT time, immediately after the RC cutoff. Great job, brain!

Here's the realization. For the Trash, we need to keep track of two things:

- where the item actually is located in the hierarchy, and

- what collection we should look at to see what permissions apply to the item.

For example, a Card might be in the Trash, but we need to look at Collection 1234 to see that a user has permission to Write that card.

My implementation of this was to add a column,
`trashed_from_collection_id`, so that we could move a Card or a Dashboard to a new `collection_id`, but keep track of the permissions we actually needed to check.

So:

- `collection_id` was where the item was located in the collection hierarchy, and

- `trashed_from_collection_id` was where we needed to look to check permissions.

Today I had the realization that it's much, much more important to get PERMISSIONS right than to get collection hierarchy right. Like if we mess up and show something as in the Trash when it's not in the Trash, or show something in the wrong Collection - that's not great, sure. But if we mess up and show a Card when we shouldn't, or show a Dashboard when we shouldn't, that's Super Duper Bad.

So the problem with my initial implementation was that we needed to change everywhere that checked permissions, to make sure they checked BOTH `trashed_from_collection_id` and `collection_id` as appropriate.

So... there's a much better solution. Instead of adding a column to represent the *permissions* that we should apply to the dashboard or card, add a column to represent the *location in the hierarchy* that should apply to the dashboard or the card.

We can simplify further: the *only time* we want to display something in a different place in the hierarchy than usual is when it was put directly into the trash. If you trash a dashboard as a part of a collection, then we should display it in that collection just like normal.

So, we can do the following:

- add a `trashed_directly` column to Cards and Dashboards, representing whether they should be displayed in the Trash instead of their actual parent collection

- use the `collection_id` column of Cards and Dashboards without modification to represent permissions.

There's one main downside of this approach. If you trash a dashboard, and then delete the collection that the dashboard was originally in, what do we do with `dashboard.collection_id`?

- we have to change it, because it's a foreign key

- we can't set it to null, because that represents the root collection

In this initial implementation, I've just cascaded the delete: if you delete a dashboard and then delete a collection, the dashboard will be deleted. This is not ideal. I'm not totally sure what we should do in this situation.

Fixes #43494

johnswanson committed:
Improve the Trash

Ok, so I had a realization at the PERFECT time, immediately after the RC
cutoff. Great job, brain!

Here's the realization. For the Trash, we need to keep track of two
things:

- where the item actually is located in the hierarchy, and

- what collection we should look at to see what permissions apply to the
item.

For example, a Card might be in the Trash, but we need to look at
Collection 1234 to see that a user has permission to Write that card.

My implementation of this was to add a column,
`trashed_from_collection_id`, so that we could move a Card or a
Dashboard to a new `collection_id`, but keep track of the permissions we
actually needed to check.

So:

- `collection_id` was where the item was located in the collection
hierarchy, and

- `trashed_from_collection_id` was where we needed to look to check
permissions.

Today I had the realization that it's much, much more important to get
PERMISSIONS right than to get collection hierarchy right. Like if we
mess up and show something as in the Trash when it's not in the Trash,
or show something in the wrong Collection - that's not great, sure. But
if we mess up and show a Card when we shouldn't, or show a Dashboard
when we shouldn't, that's Super Duper Bad.

So the problem with my initial implementation was that we needed to
change everywhere that checked permissions, to make sure they checked
BOTH `trashed_from_collection_id` and `collection_id` as appropriate.

So... there's a much better solution. Instead of adding a column to
represent the *permissions* that we should apply to the dashboard or
card, add a column to represent the *location in the hierarchy* that
should apply to the dashboard or the card.

We can simplify further: the *only time* we want to display something in
a different place in the hierarchy than usual is when it was put
directly into the trash. If you trash a dashboard as a part of a
collection, then we should display it in that collection just like
normal.

So, we can do the following:

- add a `trashed_directly` column to Cards and Dashboards, representing
whether they should be displayed in the Trash instead of their actual
parent collection

- use the `collection_id` column of Cards and Dashboards without
modification to represent permissions.

There's one main downside of this approach. If you trash a dashboard,
and then delete the collection that the dashboard was originally in,
what do we do with `dashboard.collection_id`?

- we have to change it, because it's a foreign key

- we can't set it to null, because that represents the root collection

In this initial implementation, I've just cascaded the delete: if you
delete a dashboard and then delete a collection, the dashboard will be
deleted. This is not ideal. I'm not totally sure what we should do in
this situation.

johnswanson committed:
Rip out all the `trashed_from_collection_id`

johnswanson committed:
Migration to delete trashed_from_collection_id

johnswanson committed:
fixes

johnswanson committed:
don't move collections

And don't allow deleting collections

johnswanson committed:
only show cards/dashboards with write perms

johnswanson committed:
Show the correct archived/unarchived branch

johnswanson committed:
some cleanup

johnswanson commented on src/metabase/models/collection.clj:
```diff
@@ -382,7 +386,12 @@
-                              (format "%%/%s/" (str parent-id)))]
+                              (format "%%/%s/" (str parent-id)))
+        collection-ids      (if (= collection-ids :all)
+                              (t2/select-pks-set :model/Collection :archived (or
```
There was an existing bug here. When looking for collection children, we didn't care at all about `:archived`. So with collections like `/A/B/C` where B is archived, we'd show A as having no effective children.

I'm a little concerned about performance here and still thinking about whether we can help that.

johnswanson committed:
add a todo for tomorrow

johnswanson committed:
Fix for yesterday's TODO

johnswanson committed:
more wip

johnswanson committed:
refactor

johnswanson committed:
memoize collection info

johnswanson committed:
move around memoization a bit

johnswanson committed:
fix schema migration test

johnswanson committed:
oops, delete server.middleware.memo

johnswanson committed:
Use a migration script (postgres only for now)

johnswanson committed:
Fix some tests

johnswanson committed:
remove n+1 queries in `collection-can-restore`

johnswanson committed:
fix test

johnswanson committed:
fix more tests, and x-db migration script

johnswanson committed:
fix h2 rollback

johnswanson committed:
fix mysql/mariadb migration

johnswanson committed:
lint

johnswanson committed:
fix some mariadb/mysql tests

johnswanson committed:
fix h2 rollback

johnswanson committed:
Fix mysql rollback

johnswanson committed:
Fix Postgres migration

johnswanson committed:
"Real" `trash_operation_id` UUIDs from migration

johnswanson committed:
fix mariadb migration

johnswanson committed:
Separate MySQL/MariaDB migrations

johnswanson committed:
trashed directly bit->boolean

johnswanson committed:
Remove `trashed_from_*` from migrations

johnswanson committed:
Rename `api/updates-with-trashed-directly`

Previously named `move-on-archive-or-unarchive`, which was no longer
accurate since we're not moving anything!

johnswanson committed:
Add `can_delete`

johnswanson committed:
Delete test of deleted code

johnswanson committed:
Can't move anything to the Trash

The Trash exists as a real collection so that we can list items in and
such without needing special cases. But we don't want anything to
_actually_ get moved to it.

sloansparger committed:
integrates can_delete flag on FE

rafpaf commented on frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:
```diff
@@ -52,19 +55,26 @@ export const ArchivedEntityBanner = ({
-              </BannerButton>
+              {canWrite && (
+                <BannerButton iconName="move" onClick={() => setModal("move")}>
+                  {t`Move`}
```
I know this line was not added in this PR, but this could use some `ttag.c` love to clarify that it's a verb not a noun

rafpaf reviewed (approved):
FE side looks good

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -379,31 +426,46 @@
+       ;; if the collection-ids are empty, the whole into turns into nil and we have a dangling [:and] clause in query.
+       ;; the (1 = 1) is to prevent this
+       [:and [:= [:inline 1] [:inline 1]]]
+       (if (= collection-ids :all)
```
This is never true, right? Because we already check if `collection-ids` is `:all` on L439 and then shadow the `collection-ids` var with the results of a t2 call

johnswanson replied:
Yep, good call - thanks.

noahmoss reviewed (commented):
Went through roughly half, will continue tomorrow!

johnswanson commented on resources/migrations/trash/h2.sql:
```diff
@@ -0,0 +1,101 @@
+-- DASHBOARDS
+-- Set `trashed_directly`.
+UPDATE report_dashboard
+SET trashed_directly = COALESCE(
+  -- If the collection is archived then it was *not* trashed directly
+  (SELECT NOT collection.archived FROM collection WHERE collection.id = report_dashboard.collection_id),
+  false
+  )
+WHERE archived = true;
```
Yes, that's right. One thing to note is that we're doing a kind of "best effort" thing here - we can't really know for sure whether something was trashed directly or not. For example, maybe someone trashed a dashboard, then later on trashed the collection it used to be in. We'll mark it as `trashed_directly: false` when it wasn't.

I think that's ok/inevitable, but just wanted to call it out. The knowledge of what was `trashed_directly` has been lost because we didn't record it at the moment it happened. We just do the best we can to do something reasonable.

johnswanson commented on src/metabase/api/card.clj:
```diff
@@ -189,11 +189,16 @@
         (cond->                                             ; card
           (card/model? card) (t2/hydrate :persisted)))))
 
+(defn- fix-collection-id [card]
```
Good call, I'll rename this.

johnswanson replied:
Renamed this to `api.common/present-in-trash-if-archived-directly` (the column was also renamed from `trashed_directly` to `archived_directly`)

johnswanson commented on src/metabase/api/collection.clj:
```diff
@@ -101,7 +101,14 @@
-                        (collection/permissions-set->visible-collection-ids permissions-set))]
+                        (collection/permissions-set->visible-collection-ids
+                         permissions-set
+                         {:include-archived (if archived
+                                              :only
+                                              :exclude)
```
Yeah :/

Initially I wanted to rename everything to do with `archive` to `trash`. However, I didn't want to hassle with additional migrations and the code churn required to actually change the column names (or the API) over. So that inconsistency is just there - although honestly, maybe it would be better to just use the word "Trash" on the frontend, and have the backend consistently refer to "The archive" or "archived" items. 🤔 

Then, in this case, it's worse, because I let the inconsistent column name bleed into more code. So really, I suppose this should be `include-trashed` rather than `include-archived`... even though it's checking the `:archived` column? 😭 

To answer your question about why we need both `include-archived` and `include-trash?` - `include-trash?` is whether we should include the Trash Collection itself, and `include-archived` is whether we should include things that are "in" the trash.

noahmoss replied:
I'm personally in favor of consistent naming even if it requires migrations/a little more code churn. Otherwise we built up a bank of implicit historical knowledge you need to figure out to approach the codebase. 

But if it's too much to do as part of this PR, maybe open a tech debt issue for it? At least it'll be documented and could be a good onboarding task in the future.

For the naming of these flags, maybe `include-archived-items?` and `include-trash-collection?`?

johnswanson replied:
Renamed to `:include-archived-items` and `:include-trash-collection?`, and replaced `trashed` with `archived` everywhere, including in migrations (e.g. `trashed_directly` => `archived_directly` and `trash_operation_id` => `archive_operation_id`)

johnswanson commented on src/metabase/api/collection.clj:
```diff
@@ -505,7 +526,8 @@
   (-> (t2/instance :model/Card row)
       (update :collection_preview api/bit->boolean)
       (update :archived api/bit->boolean)
-      (t2/hydrate :can_write :can_restore)
+      (update :trashed_directly api/bit->boolean)
```
Ooh, I wonder. It does seem like if JDBC interprets `bit(1)` as a boolean we should be able to do away with this.

I can check after the other PR is merged.

johnswanson replied:
Though I did add `trashed_directly` as a `${boolean.type}` in the migrations. So something odd is going on. 🤔

johnswanson commented on src/metabase/models/collection.clj:
```diff
@@ -306,41 +302,92 @@
 (def ^:private VisibleCollections
   "Includes the possible values for visible collections, either `:all` or a set of ids, possibly including `\"root\"` to
```
Good call, thanks - will take that out.

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -731,69 +799,101 @@
+    @api/*current-user-permissions-set*
+    (perms-for-archiving collection)))
+  (let [trash-operation-id      (str (random-uuid))
+        affected-collection-ids (cons (u/the-id collection)
+                                      (collection->descendant-ids collection
+                                                                  :archived [:not= true]))]
```
Should this be in the transaction?

johnswanson replied:
Yeah, that makes sense - done!

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -731,69 +799,101 @@
+        :where  [:and
+                 [:like :location (str (children-location collection) "%")]
+                 [:not :archived]]})
+      (doseq [model [:model/NativeQuerySnippet :model/Pulse :model/Timeline]]
+        (t2/update! model {:collection_id [:in affected-collection-ids]}
+                    {:archived true}))
+      (doseq [model [:model/Card :model/Dashboard]]
```
There are a couple places now in this namespace where we enumerate the "models which can live in a collection" which makes me worried that it'll be easy to miss updating these if a new model is added. 

Maybe this could be pulled out as a standalone var? I kind of like the concept of `collectable-models`. Then you'd only need to hard-code Card and Dashboard in these functions as the ones that can be trashed directly (or that could be a separate var)

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -1492,68 +1572,65 @@
+                                                     @api/*current-user-permissions-set*
+                                                     (perms-for-archiving coll)))))))))
+
+(mi/define-batched-hydration-method can-delete
+  :can_delete
+  "Efficiently hydrate the `:can_delete` of a sequence of items"
+  [items]
+  (when (seq items)
+    (for [item items]
+      (assoc item :can_delete (mi/can-write? item)))))
+
+(mi/define-batched-hydration-method collection-can-delete
+  :collection/can_delete
+  "Right now, you can never permanently delete collections. This is a placeholder in case that changes."
+  [items]
+  (for [item items]
+    (assoc item :can_delete false)))
```
Why not just have the `:can_delete` hydration detect whether the item is a collection? Otherwise we're relying on the caller to use the correct hydration method

johnswanson replied:
Oh, that's much better - thank you!

noahmoss commented on src/metabase/search/impl.clj:
```diff
@@ -553,14 +545,17 @@
+                                   :name            collection_name
+                                   :authority_level collection_authority_level
+                                   :type            collection_type}
+                                  (when effective_parent
+                                    effective_parent)
```
this was my own code, but I just realized that we don't need the `when` since if `effective_parent` is nil it'll be ignored by the merge 😀 
```suggestion
                                  ;; for  non-root collections, override :collection with the values for its effective parent
                                  effective_parent
```

noahmoss reviewed (approved):
Finished going through, looks good to me overall! Most comments are minor so I'm stamping to unblock, but I'm happy to look again at any parts that change

johnswanson committed:
Better name for `fix-collection-id`

Also consolidated it into a single function. It takes the
trash-collection-id as an argument to prevent a circular dependency.

johnswanson committed:
s/trashed/archived/ on the backend

In the product, we want to move to "trashed" as the descriptor for
things that are currently-known-as-archived.

It'd be nice to switch to that on the backend as well, but that'll
require quite a lot of churn on the frontend as we'd need to change the
API (or have an awkward translation layer, where we call it `trashed` on
the backend, turn it into `archived` when sending data out via the API,
and then the FE turns around and called it "trashed" again).

For now, let's just be consistent on the backend and call it `archived`
everywhere. So:

- `archived_directly` instead of `trashed_directly`, and
- `archive_operation_id` instead of `trash_operation_id`

johnswanson committed:
Fix up a couple docstrings

johnswanson committed:
Rename `visible-collection-ids` args

for `collection/permissions-set->visible-collection-ids`:

- `:include-archived` => `:include-archived-items`

- `:include-trash?` => `:include-trash-collection`

johnswanson committed:
Update src/metabase/models/collection.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

johnswanson committed:
select affected-collection-ids in the tx

johnswanson committed:
Stop dealing with `:all` visible collection ids

This used to be a possible return value for
`permissions-set->visible-collection-ids` indicating that the user could
view all collections. Now that function always returns a set of
collection IDs (possibly including "root" for the root collection) so we
don't need to deal with that anymore.

johnswanson committed:
Update src/metabase/search/impl.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

johnswanson committed:
Don't use separate hydration keys for collections

Dispatch on the model of each item (but still keep them in batches for
efficiency.)

johnswanson committed:
vars for collectable/archived-directly models

johnswanson committed:
Round up loose `trashed_directly`s

johnswanson committed:
Fix a test

johnswanson committed:
Use `clojure.core.memoize` instead of handrolled

It's slightly different (TTL instead of exactly per-request) but that
should be fine.

johnswanson commented on src/metabase/api/common.clj:
```diff
@@ -658,21 +658,38 @@
-                            :else (:trashed_from_collection_id current-obj))
-           :trashed_from_collection_id (when (:archived obj-updates)
-                                         (:collection_id current-obj)))))
+;; ----------------------------------------- PER-REQUEST MEMOIZATION ---------------------------------------------------
```
This seems like a good idea - I've:
- [added a *request-id* dynamic var to config](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-a02bbaeb1685f3f566f2913949ecf046bf8507f757dcc6bd0359ee89bd12d0c4R163-R165)
- [bound that in a new middleware](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-90cbda8884d3b0752f52adb405bfb073b552341b2860cf210f2753c5a888186fR1-R9)
- [used that](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-ddea8b16dfd87ccdb499efcc15deca578afe6982b7bba016c75510b6bc484529R315-R328) as the cache args for `collection-id->collection`

johnswanson committed:
FE lint fixes

johnswanson committed:
:sob: e2e test fix
